### ISSUE TITLE
Merging Federation API with node API into single repo

### ISSUE THREAD

**Issue by rmanaem** [neurobagel/planning#172]
Given the majority of the code in federation API is often derived and copied over from the node API, we want to discuss whether this service can exist as a module in the node API to facilitate a more efficient development process.

TODO:

- [x] Make a list of logic that is shared in both API repos and put it in a shared document
- [ ] explain how we can build two separate docker images, one per app, from the same repo
- [ ] How do releases work if both tools are in the same repo
- [ ] If both apps share code and we factor that out, how do we handle it during the docker image build process (i.e. is the common part a dependency or just included in both builds directly)
- [ ] What tools exist to make handling monorepos easier, and what do they do for us

## Tasks

- [x] Create docker-compose that builds and then deploys both apps as they would run in production
- [x] Document how to setup dev environment for each app and run tests
- [x] Document how to run integration tests (that don't mock either app)
- [ ] Make a decision on releasing the tools using their current versions, on releasing them independently, or with a shared version

**Comment by rmanaem** [comment]
some overlap: a good deal of overlapping logic coupled with app-specific logic so it's possible to merge but requires refactoring
little overlap:  the files could be merged but will run into some conflicts (same function names with similar functionality but different logic)
no overlap: the files could be merged with almost no conflict
a lot of overlap: essentially very similar code and logic, could be merged with a small refactor

routers:
- assessment.py: some overlap
- attributes.py: only in n-api
- diagnoses.py: some overlap
- pipelines.py: some overlap
- query.py: a lot of overlap
- node.py: only in f-api
- route-factory: some overlap

other:
- crud.py: little overlap
- models.py: some overlap
- security.py: a lot of overlap
- utility.py: no overlap
- main.py: a lot of overlap

**Comment by alyssadai** [comment]
Some initial thoughts after our chat today:
- Having two separate apps in the same repo, but each potentially needing to be released independently of the other, seems it could be challenging to maintain (since we'd essentially be violating the expectation of what a repo is). My sense is that what we'd reduce for ourselves in terms of cross-repo changes, we would gain in devops work needed to make and maintain separate, automatic releases of two apps from a single repo.
  - If the goal is to still release each API separately, I think keeping them two repos with common logic factored out would be an option that improves dev experience without adding a lot of overhead for managing essentially two distinct codebases in one place
- If we want to prioritize having the two apps in one repo, I think at minimum we should release them both together (i.e., whenever one API needs to be released, both are)
  - This gives us a bit more flexibility if we decide to internally refactor the logic of the APIs, and also means we can still take advantage of GH's release system since we're staying closer to the expectation of what a repo is (e.g., v0.5.0 could be now just associated w/ two compatible Docker images, each also being v0.5.0) 

Some questions:
- From a user's perspective, assuming we do proceed with merging the API repos, how would they report bugs for just one API? Maybe we use separate issue templates for the f-API and n-API?

**Comment by surchs** [comment]
I agree with Alyssa's points. We have talked a good bit about this now, and decided to not do the API repo merge. The main reasons for not doing the merge are

- an n-API and f-API that talk to each other are most often parts of different deployments (e.g. one sits at site A and talks to the other at site B). And so we don't help ourselves by just ensuring that the most recent version of n-API and f-API can talk to each other, because across the network they are not upgraded at the same time (again, because part of different deployments)
- merging the repos makes it more involved to version n-API and f-API separately - which we need to continue doing for above reason
- there are other ways to achieve the same DX improvement of ready-to-go integration tests, development environments, and ability to stand up the other needed services during testing - but that's a separate conversation

In the process of discussing all this, we also made it a bit clearer that unlike n-API and f-API, most of our other tools actually can be considered part of the same deployments. i.e. CLI, graph data schema, graph, and n-API (arguably also annotation tool, and data dictionary schema) together all make up a "node bunndle" (name tbd), and f-API and query tool together make up a "federation bundle". Most places deploy both both "bundles" at the same time, or even as part of the same stack. But it is reasonable to ask users to upgrade everything in a "node bundle" together, at least for now.

So we'll close this issue, retire the merged repo, and discuss our versioning and DX improvements in other issues.
